R/estimate_proportion.R

```diff
@@ -32,14 +32,142 @@ prop_wilson <- function(rsp, conf_level, correct = FALSE) {
 }
 
 
+# help function for prop_strat_wilson
+strata_normal_quantile <- function(vars, weights, conf_level) {
+  summands <- weights^2 * vars
+  # Stratified quantile
+  sqrt(sum(summands)) / sum(sqrt(summands)) * qnorm((1 + conf_level) / 2)
+}
+
+#' @describeIn estimate_proportions Calculates the stratified Wilson confidence
+#'   interval for unequal proportions as described in [Yan and Su (2010)].
+#'
+#' @param strata (`factor`)\cr
+#'   with one level per stratum and same length as `rsp`.
+#' @param weights weights for each level of the strata. If missing, they are
+#'   estimated using the iterative algorithm proposed in [Yan and Su (2010)]
+#'   that minimizes the weighted squared length of the confidence interval.
+#' @param max_nit maximum number of iterations for the iterative procedure used
+#'   to find estimates of optimal weights.
+#' @param correct `logical`\cr
+#'   include the continuity correction. For further information, see for example
+#'   [stats::prop.test()].
+#'
+#' @examples
+#' # Stratified Wilson confidence interval with unequal probabilities
+#'
+#' set.seed(1)
+#' rsp <- sample(c(TRUE, FALSE), 100, TRUE)
+#' strata_data <- data.frame(
+#'   "f1" = sample(c("a", "b"), 100, TRUE),
+#'   "f2" = sample(c("x", "y", "z"), 100, TRUE),
+#'   stringsAsFactors = TRUE
+#' )
+#' strata <- interaction(strata_data)
+#' n_ws <- ncol(table(rsp, strata)) # Number of weights
+#'
+#' prop_strat_wilson(
+#'   rsp = rsp, strata = strata,
+#'   conf_level = 0.90
+#' )
+#'
+#' prop_strat_wilson(
+#'   rsp = rsp, strata = strata,
+#'   weights = rep(1 / n_ws, n_ws), # Not automatic setting of weights
+#'   conf_level = 0.90
+#' )
+#'
+#' @export
+prop_strat_wilson <- function(rsp, strata, weights = NULL, conf_level = 0.95, max_nit = NULL, correct = FALSE) {
+
+  # Checking the inputs
+  checkmate::assert_logical(rsp, any.missing = FALSE)
+  checkmate::assert_factor(strata, len = length(rsp))
+  tern:::assert_proportion_value(conf_level)
+
+  tbl <- table(rsp, strata)
+  n_ws <- ncol(tbl) # Number of centers
+
+  # Checking the weights and maximum number of iterations
```

Comment:
```suggestion
  # Checking the weights and maximum number of iterations.
```
(sentence style for comments)